server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconcilerTests.java

```diff
@@ -736,6 +736,79 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
         }
     }
 
+    public void testUnassignedPrimariesThrottlingAndFallback() {
+        // we fall back to trying all nodes if an unassigned primary cannot be assigned to a desired node, but only if the desired nodes
+        // aren't just throttled
+
+        final var discoveryNodes = discoveryNodes(2);
+        final var metadata = Metadata.builder();
+        final var routingTable = RoutingTable.builder();
+
+        final var indexMetadata0 = randomPriorityIndex("index-0", 2, 0);
+        metadata.put(indexMetadata0, true);
+        routingTable.addAsNew(indexMetadata0);
+
+        final var clusterState = ClusterState.builder(ClusterName.DEFAULT)
+            .nodes(discoveryNodes)
+            .metadata(metadata)
+            .routingTable(routingTable)
+            .build();
+
+        final var settings = throttleSettings();
+        final var clusterSettings = new ClusterSettings(settings, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
+
+        final var desiredBalance = desiredBalance(clusterState, (shardId, nodeId) -> nodeId.equals("node-0"));
+        final var allocationFilter = new AtomicReference<BiPredicate<Integer, String>>((shardId, nodeId) -> shardId == 0);
+
+        final var allocationService = createTestAllocationService(
+            routingAllocation -> reconcile(routingAllocation, desiredBalance),
+            new SameShardAllocationDecider(settings, clusterSettings),
+            new ReplicaAfterPrimaryActiveAllocationDecider(),
+            new ThrottlingAllocationDecider(settings, clusterSettings),
+            new AllocationDecider() {
+                @Override
+                public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+                    return allocationFilter.get().test(shardRouting.getId(), node.nodeId()) ? Decision.YES : Decision.NO;
+                }
+            }
+        );
+
+        final var unused = ActionListener.<Void>noop();
+
+        // first assign the primary of [index-0][0] (no other shards may be allocated due to allocation filter)
+        final var stateWithOneInitializingPrimary = allocationService.reroute(clusterState, "test", unused);
```

Comment:
NIT: set the initial value `allocationFilter` here so that it is closer to the place where it is used.

Reply:
Good call, see ff41a13.